---
#title: "Cebrian-Silla, Nascimento, Redmond, Mansky et al. 2021 - Analysis: A Cells"
output: rmarkdown::github_document
---
#setup
```{r setup}
library(knitr)
library(future)
library(ggplot2) 
library(reticulate) 
library(Seurat)
library(R.utils)
library(dplyr)
library(tibble)
library(viridis)
library(patchwork)
library(dendextend)
library(stats)
library(remotes)
library(readr)
library(biomaRt)
library(enrichplot)
library(scales)
library(tidyr)
library(scico)
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

#Loading the annotated Seurat Object
```{r}
exp <- readRDS("../Files/seurat/GSM5039270_scSeq.rds")
Idents(exp) <- "integrated_snn_res.1.5"
DimPlot(exp, label = T) + coord_fixed() + NoLegend()

exp.lin <- subset(exp, idents = c("14", "5", "22", "13", "12", "10", "16", "17", "8", "15", "1", "6", "4", "0"))
DimPlot(exp.lin, label = T) + coord_fixed() + NoLegend()

exp.a <- subset(exp, idents = c("0", "1", "4", "6", "15"))
DimPlot(exp.lin, label = T) + coord_fixed() + NoLegend()
```

#Subsetting A cells
```{r}
a.cells.palette <- c("#671313", "#9B1C1C", "#D93030", "#DF5353", "#EC9898")

a.cells <- WhichCells(exp, idents = c(15,6,1,4,0))
exp@meta.data[a.cells,"acells"] <- exp@meta.data[a.cells,"integrated_snn_res.1.5"]

DimPlot(exp, group.by = "acells", label=T, shuffle = T, cols = a.cells.palette ) + 
  coord_fixed() + 
  xlim(-6, 7) + ylim(-15, 8) + 
  NoAxes() + 
  labs(title = "A Cells")
ggsave

DimPlot(exp, cells = a.cells, label=T, shuffle = T, cols = a.cells.palette ) + 
  xlim(-4.5, 3.5) + ylim(-14, -2) + 
  coord_fixed() + 
  NoAxes() + 
  labs(title = "A Cells Subset")

DimPlot(exp, cells = a.cells, label=F, group.by = "Phase") + 
  xlim(-4.5, 3.5) + ylim(-14, -2) + 
  coord_fixed() + 
  NoAxes() + 
  labs(title = "A Cells - Cell cycle labeling")
```

##Checking gene expression

#exp.a Markers
##FindAllMarkers in exp.a
```{r}
a.markers <- FindAllMarkers(exp.a, only.pos = T)

cutoff <- 0.5 

cluster15.markers <- a.markers %>% filter(cluster==15 & avg_logFC >0 & pct.2 < cutoff) %>% pull(gene)
cluster6.markers <- a.markers %>% filter(cluster==6 & avg_logFC >0 & pct.2 < cutoff) %>% pull(gene)
cluster4.markers <- a.markers %>% filter(cluster==4 & avg_logFC >0 & pct.2 < cutoff) %>% pull(gene)
cluster1.markers <- a.markers %>% filter(cluster==1 & avg_logFC >0 & pct.2 < cutoff) %>% pull(gene)
cluster0.markers <- a.markers %>% filter(cluster==0 & avg_logFC >0 & pct.2 < cutoff) %>% pull(gene)

write.table(cluster15.markers, quote = F, row.names = F, col.names = F, "cluster15.strictmarkers.txt", sep=",")
write.table(cluster6.markers, quote = F, row.names = F, col.names = F, "cluster6.strictmarkers.txt", sep=",")
write.table(cluster4.markers, quote = F, row.names = F, col.names = F, "cluster4.strictmarkers.txt", sep=",")
write.table(cluster1.markers, quote = F, row.names = F, col.names = F, "cluster1.strictmarkers.txt", sep=",")
write.table(cluster0.markers, quote = F, row.names = F, col.names = F, "cluster0.strictmarkers.txt", sep=",")

#Judging by the number of markers, clusters 0 (dorsal), 1 (ventral), and 6 (proliferative) are very unique. Cluster 4 seems like a transitional identity

#String Analysis:
  #Cluster1/Ventral: 
    # 0.2 cutoff: https://version-11-0.string-db.org/cgi/network.pl?networkId=WWV16gHY9tKm #20 genes
    # no cutoff: https://version-11-0.string-db.org/cgi/network.pl?networkId=rTkKFUYrPbYl #121 genes

  #Cluster0/Dorsal: 
    #https://version-11-0.string-db.org/cgi/network.pl?networkId=8WC9g7rgGcxQ #10 genes
    #https://version-11-0.string-db.org/cgi/network.pl?networkId=OHEYxcivrclk #60 genes

  #Cluster6: Early neuroblasts
```

####Plots
```{r}
#cluster 15
p1 <- FeaturePlot(exp.a, features = cluster15.markers[1:9], combine = FALSE, pt.size = 0.6, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() + xlim(-4.5, 3.5) + ylim(-14, -5) + fix.sc + simple)

wrap_plots(Plot1, ncol = 3) + 
  plot_annotation(
    title = "Top cluster 15 markers",
    subtitle = "Identified using FindAllMarkers() in exp.a",
    caption = "Code available on A.cells.Rmd")



#cluster 6
p1 <- FeaturePlot(exp.a, features = cluster6.markers[1:9], combine = FALSE, pt.size = 0.6, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() + xlim(-4.5, 3.5) + ylim(-14, -5) + fix.sc + simple)

wrap_plots(Plot1, ncol = 3) + 
  plot_annotation(
    title = "Top cluster 6 markers",
    subtitle = "Identified using FindAllMarkers() in exp.a",
    caption = "Code available on A.cells.Rmd")



#cluster 4
p1 <- FeaturePlot(exp.a, features = cluster4.markers[1:9], combine = FALSE, pt.size = 0.6, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() + xlim(-4.5, 3.5) + ylim(-14, -5) +  fix.sc + simple)

wrap_plots(Plot1, ncol = 3) + 
  plot_annotation(
    title = "Top cluster 4 markers",
    subtitle = "Identified using FindAllMarkers() in exp.a",
    caption = "Code available on A.cells.Rmd")



#cluster 1
p1 <- FeaturePlot(exp.a, features = cluster1.markers[1:9], combine = FALSE, pt.size = 0.6, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() + xlim(-4.5, 3.5) + ylim(-14, -5) +  fix.sc + simple)

wrap_plots(Plot1, ncol = 3) + 
  plot_annotation(
    title = "Top cluster 1 markers",
    subtitle = "Identified using FindAllMarkers() in exp.a",
    caption = "Code available on A.cells.Rmd")



#cluster 0
p1 <- FeaturePlot(exp.a, features = cluster0.markers[1:9], combine = FALSE, pt.size = 0.6, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() + xlim(-4.5, 3.5) + ylim(-14, -5) +  fix.sc + simple)

wrap_plots(Plot1, ncol = 3) + 
  plot_annotation(
    title = "Top cluster 0 markers",
    subtitle = "Identified using FindAllMarkers() in exp.a",
    caption = "Code available on A.cells.Rmd")

```
###Cluster 1 vs Cluster 0
```{r}
Idents(exp.a) <- "integrated_snn_res.1.5"
vxd.a.markers <- FindMarkers(exp.a, ident.1 = 1, ident.2 = 0)

v.a.markers <- rownames(vxd.a.markers %>% filter(avg_logFC >0 & pct.2 < 0.5))
d.a.markers <- rownames(vxd.a.markers %>% filter(avg_logFC <0 & pct.1 < 0.5))
v.a.markers
d.a.markers
```

####Plots
```{r}
#cluster 1
p1 <- FeaturePlot(exp.a, features = v.a.markers[1:9], combine = FALSE, pt.size = 0.6, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() + xlim(-4.5, 3.5) + ylim(-14, -5) +  fix.sc + simple)

wrap_plots(Plot1, nrow = 3) + 
  plot_annotation(
    title = "Top cluster 1 markers",
    subtitle = "Identified comparing clusters 1 and 0",
    caption = paste0("Generated on ", format(Sys.time(), "%D"), ". Code available on A.cells.Rmd"))

VlnPlot(exp.a, features = v.a.markers[1:20], stack = T, flip = T, idents = c("1","0")) + NoLegend()




#cluster 0
p1 <- FeaturePlot(exp.a, features = d.a.markers[1:9], combine = FALSE, pt.size = 0.6, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() + xlim(-4.5, 3.5) + ylim(-14, -5) +  fix.sc + simple)

wrap_plots(Plot1, nrow = 3) + 
  plot_annotation(
    title = "Top cluster 0 markers",
    subtitle = "Identified comparing clusters 1 and 0",
    caption = paste0("Generated on ", format(Sys.time(), "%D"), ". Code available on A.cells.Rmd"))

VlnPlot(exp.a, features = d.a.markers[1:20], stack = T, flip = T, idents = c("1","0")) + NoLegend()
```


#Assigning A cell identities
```{r}
Idents(exp.a) <- "integrated_snn_res.1.5"
v.acells <- WhichCells(exp.a, idents = 1)
d.acells <- WhichCells(exp.a, idents = 0)
nb.acells <- WhichCells(exp.a, idents = c(6, 15))
inter.acells <- WhichCells(exp.a, idents = 4)


exp.a@meta.data[c(v.acells, inter.acells, d.acells),  "cellid"] <- "Young Neurons"
exp.a@meta.data[nb.acells,  "cellid"] <- "Neuroblasts"


Idents(exp.a) <- "cellid"
DimPlot(exp.a, group.by = "cellid", label = T) + coord_fixed() + xlim(-4.5, 3.5) + ylim(-14, -2) + simple

a.id.markers <- FindAllMarkers(exp.a, only.pos = T)
a.id.markers
```


#Heatmap + Dendrogram
```{r}
Idents(exp.a) <- "integrated_snn_res.1.5"
exp.a.heatmap <- SCTransform(exp.a, new.assay.name = "SCT2", variable.features.n = 20681, vars.to.regress = "percent.mt")
exp.a.heatmap <- BuildClusterTree(exp.a.heatmap, reorder=T)

#Dendrogram
dend1 <- Tool(exp.a.heatmap, 'BuildClusterTree')
dend1
dend1 <- as.dendrogram(dend1)
ggplot(dend1)


#Heatmap
heatmap<- DoHeatmap(subset(exp.a.heatmap, downsample = 200), features = c(cluster1.markers[1:5], cluster0.markers[1:5], cluster4.markers[1:5], cluster6.markers[1:5],cluster15.markers[1:5]), size = 3)
heatmap
```

#Feature plots for regional markers
```{r}
FeaturePlot(exp.a, features = "Slit2", order = T, pt.size = 0.7) + 
  scale_color_viridis(alpha = 1, option = "A") + 
  coord_fixed() +
  xlim(-4.5, 3.5) + 
  ylim(-14, -2) +
  simple
ggsave(filename = "slit2.png", height = 2, width = 2, scale = 2)

FeaturePlot(exp.a, features = "Vax1", order = T, pt.size = 0.7) + 
  scale_color_viridis(alpha = 1, option = "A") + 
  coord_fixed() +
  xlim(-4.5, 3.5) + 
  ylim(-14, -2) +
  simple
ggsave(filename = "vax1.png", height = 2, width = 2, scale = 2)

FeaturePlot(exp.a, features = "Pax6", order = T, pt.size = 0.7) + 
  scale_color_viridis(alpha = 1, option = "A") + 
  coord_fixed() +
  xlim(-4.5, 3.5) + 
  ylim(-14, -2) +
  simple
ggsave(filename = "pax6.png", height = 2, width = 2, scale = 2)

FeaturePlot(exp.a, features = "Rlbp1", order = T, pt.size = 0.7) + 
  scale_color_viridis(alpha = 1, option = "A") + 
  coord_fixed() +
  xlim(-4.5, 3.5) + 
  ylim(-14, -2) +
  NoAxes()
ggsave(filename = "rlbp1.png", height = 2, width = 2, scale = 2)

```

#Looking for DE genes (dataset-wide) in A cell clusters:
```{r}
#Find All Markers
all.markers <- FindAllMarkers(exp, only.pos = T)

##Listing markers for each cluster
cluster15.allmarkers <- all.markers %>% filter(cluster==15) %>% pull(gene)
cluster6.allmarkers <- all.markers %>% filter(cluster==6) %>% pull(gene)
cluster4.allmarkers <- all.markers %>% filter(cluster==4) %>% pull(gene)
cluster1.allmarkers <- all.markers %>% filter(cluster==1) %>% pull(gene)
cluster0.allmarkers <- all.markers %>% filter(cluster==0) %>% pull(gene)

write.table(cluster15.allmarkers, quote = F, row.names = F, col.names = F, "cluster15.allmarkers.txt", sep=",")
write.table(cluster6.allmarkers, quote = F, row.names = F, col.names = F, "cluster6.allmarkers.txt", sep=",")
write.table(cluster4.allmarkers, quote = F, row.names = F, col.names = F, "cluster4.allmarkers.txt", sep=",")
write.table(cluster1.allmarkers, quote = F, row.names = F, col.names = F, "cluster1.allmarkers.txt", sep=",")
write.table(cluster0.allmarkers, quote = F, row.names = F, col.names = F, "cluster0.allmarkers.txt", sep=",")
```

##Plotting DE genes in A cell clusters
```{r}
#cluster15
p1 <- FeaturePlot(exp.a, features = cluster15.allmarkers[1:9], combine = FALSE, pt.size = 0.2, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() +  fix.sc + simple + xlim(-6, 7) + ylim(-15, 8))

wrap_plots(Plot1, nrow = 3) + 
  plot_annotation(
    title = "Top cluster 15 markers",
    subtitle = "Identified using FindAllMarkers() in exp",
    caption = "Code available on A.cells.Rmd")


#cluster6
p1 <- FeaturePlot(exp.a, features = cluster6.allmarkers[1:9], combine = FALSE, pt.size = 0.2, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() +  fix.sc + simple + xlim(-6, 7) + ylim(-15, 8))

wrap_plots(Plot1, nrow = 3) + 
  plot_annotation(
    title = "Top cluster 6 markers",
    subtitle = "Identified using FindAllMarkers() in exp",
    caption = "Code available on A.cells.Rmd")


#cluster4
p1 <- FeaturePlot(exp.a, features = cluster4.allmarkers[1:9], combine = FALSE, pt.size = 0.2, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() +  fix.sc + simple)

wrap_plots(Plot1, nrow = 3) + 
  plot_annotation(
    title = "Top cluster 4 markers",
    subtitle = "Identified using FindAllMarkers() in exp",
    caption = "Code available on A.cells.Rmd")


#cluster1
p1 <- FeaturePlot(exp.a, features = cluster1.allmarkers[1:9], combine = FALSE, pt.size = 0.2, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() +  fix.sc + simple)

wrap_plots(Plot1, nrow = 3) + 
  plot_annotation(
    title = "Top cluster 1 markers",
    subtitle = "Identified using FindAllMarkers() in exp",
    caption = "Code available on A.cells.Rmd")


#cluster0
p1 <- FeaturePlot(exp.a, features = cluster0.allmarkers[1:9], combine = FALSE, pt.size = 0.2, order = T)
fix.sc <- scale_color_viridis(alpha = 1, option = "A") 
simple <- NoAxes() + NoLegend() 
Plot1 <- lapply(p1, function (x) x + coord_fixed() +  fix.sc + simple)

wrap_plots(Plot1, nrow = 3) + 
  plot_annotation(
    title = "Top cluster 0 markers",
    subtitle = "Identified using FindAllMarkers() in exp",
    caption = "Code available on A.cells.Rmd")

```

##GO analysis
```{r}
#GO analysis was performed by importing the markers from each cluster into pantherdb, using the .txt files created in this notebook.

#In Pantherdb, exported all GO terms for DE genes in each cluster as a .txt file named "clusterX.allmarkers.txt" to a subfolder called "GO"
#A for loop to read all .txt files exported from pantherdb and create tables names "cluster#.go"
clusters <- c(0,1,4,6,15)
for (i in clusters) {
  assign(paste0("cluster",i,".go"), read_delim(paste0("GO/cluster", i,  ".allmarkers.txt"), 
    "\t", escape_double = FALSE, col_names = c("go", "ref", "obs", "expected", "pos.neg", "fold.enrich", "raw.p.value", "fdr"), 
    trim_ws = TRUE, skip = 12))
} 
clusterorder <- c("Cluster 1", "Cluster 0", "Cluster 4", "Cluster 6", "Cluster 15")

#create a dataframe with all categories and clusters:
go.cat.n <- nrow(cluster1.go)
go.data <- do.call("rbind", list(cluster1.go, cluster0.go, cluster4.go, cluster6.go, cluster15.go))
go.data[1:go.cat.n,"cluster"] <- "Cluster 1"
go.data[(1+go.cat.n):(2*go.cat.n),"cluster"] <- "Cluster 0"
go.data[(1+(2*go.cat.n)):(3*go.cat.n),"cluster"] <- "Cluster 4"
go.data[(1+(3*go.cat.n)):(4*go.cat.n),"cluster"] <- "Cluster 6"
go.data[(1+(4*go.cat.n)):(5*go.cat.n),"cluster"] <- "Cluster 15"
go.data$cluster <- factor(go.data$cluster, levels = clusterorder)

saveRDS(go.data, file = "A.cells.go.data.rds")
                   
#plotting barplot
df <- go.data %>% filter(raw.p.value < 0.01 & fdr <0.05 & cluster =="Cluster 0") %>% top_n(20, fold.enrich)


df$go <- factor(df$go, levels= rev(unique(as.character(df$go))))
#df <- transform(df, go=reorder(variable, -value)) #to reorder the rows according to another value

#single cluster
ggplot(df, aes(x= fold.enrich, y = go, fill = fdr)) + 
  scale_fill_gradient() + 
  geom_bar(stat = "identity") + 
  scale_fill_continuous(high = "#fff7ec", low = "#7f0000") + 
  theme_minimal() + 
  theme(panel.grid.major.y = element_blank()) + 
  labs(title = "Cluster 0")

#allclusters wrapped
##Finding the top N categories enriched in each cluster, filtering for p value < 0.01 and FDR <0.05:
go.cats <- unique(go.data %>% filter(raw.p.value < 0.01 & fdr <0.05) %>% group_by(cluster) %>% slice_max(n = 10, order_by = fold.enrich, with_ties = F) %>% pull(go))
go.cats.curated <- c("mitotic DNA replication initiation (GO:1902975)",
                     "DNA strand elongation involved in DNA replication (GO:0006271)",
                     "DNA replication, synthesis of RNA primer (GO:0006269)",
                     "mitotic DNA replication (GO:1902969)",
                     "positive regulation of DNA-directed DNA polymerase activity (GO: 1900264)",
                     "regulation of postsynaptic density protein 95 clustering (GO:1902897)",
                      "nucleosome disassembly (GO:0006337)",
                     "ephrin receptor signaling pathway (GO:0048013)",
                     "commitment of multipotent stem cells to neuronal lineage in forebrain (GO:0021898)",
                     "axon guidance (GO:0007411)",
                     "chemotaxis (GO:0006935)",
                    "regulation of postsynaptic density organization (GO:1905874)",
                     "neural precursor cell proliferation (GO:0061351)",
                     "olfactory bulb interneuron differentiation (GO:0021889)",
                     "regulation of synaptic vesicle clustering (GO:2000807)",
                     "regulation of negative chemotaxis (GO:0050923)",
                     "neuroligin clustering involved in postsynaptic membrane assembly (GO:0097118)",
                     "positive regulation of dendritic spine morphogenesis (GO:0061003)",
                     "cerebral cortex radial glia guided migration (GO:0021801)",
                     "anterograde dendritic transport (GO:0098937)",
                     "modulation of inhibitory postsynaptic potential (GO:0098828)",
                     "spontaneous synaptic transmission (GO:0098814)",
                     "spontaneous neurotransmitter secretion (GO:0061669)")

#creating dataframe to use in plots:
df.all<- go.data %>% filter(go %in% go.cats.curated)
df.all <- df.all %>% mutate(fold.enrich = replace_na(fold.enrich, 0))
df.all <- df.all %>% mutate(sig = ifelse(fdr < 0.05, T, F))
df.all$cluster <- factor(df.all$cluster, levels = clusterorder)
df.all$go <- factor(df.all$go, levels = go.cats.curated)

#Heatmap
ggplot(df.all %>% mutate(fold.enrich = replace_na(fold.enrich, 0)), aes(cluster, go)) + geom_tile(aes(fill = fold.enrich)) + scale_fill_viridis() + theme_minimal()

#Dotplot
ggplot(df.all, aes(cluster, go)) + 
  geom_point(aes(col = sig, size = fold.enrich)) + 
  scale_color_manual(name = NULL, labels = c("FDR p > 0.05", "FDR p < 0.05"), values = viridis(2, direction = -1)) + 
  scale_size(name = "Fold enrichment") +
  scale_x_discrete(labels=c("Cluster 1" = "1", "Cluster 0" = "0", "Cluster 4" = "4", "Cluster 6" = "6", "Cluster 15" = "15")) +
  theme_minimal() + 
  theme(panel.grid.major = element_blank(), axis.title = element_blank(), legend.position = "bottom", legend.direction = "horizontal")
```
